{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "MODEL_NUM = \"EXAMPLE\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.linear_model import ElasticNetCV\n",
    "from scipy.stats import pearsonr\n",
    "from multiprocessing import Pool, cpu_count\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_annotations = pd.read_csv(\"../example_data/snp_annotations.txt\")\n",
    "gene_annotations = pd.read_csv(\"../example_data/gene_annotations.txt\", index_col=0)\n",
    "dosages = pd.read_csv(\"../example_data/dosages.txt\", index_col=0)\n",
    "clusterings = pd.read_csv(\"../example_data/clusterings.txt\", index_col=0)\n",
    "expression_phenotypes = pd.read_csv(\"../example_data/expression_phenotypes.txt\", index_col=0)\n",
    "tissue = \"example\"\n",
    "\n",
    "# Differentiate haplotypes in index labels\n",
    "clusterings.index = [start_var+\"_h\"+str(i%2) for i,start_var in enumerate(clusterings.index)]\n",
    "\n",
    "# Split SNP annotations by chromosome for faster searching\n",
    "snp_annotations_by_chr = {chr: snp_annotations[snp_annotations[\"chr\"] == chr] for chr in set(snp_annotations[\"chr\"])}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Model parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "SEED = 42\n",
    "\n",
    "# Model parameters\n",
    "ciswindow_size = 1e6\n",
    "n_folds = 10\n",
    "n_alphas = 10\n",
    "tol = 1e-7"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define model and evaluation environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_haplotype_cluster_dummies(X):\n",
    "    def _cat_to_vec(category, category_to_int_dict):\n",
    "        categories = category_to_int_dict.keys()\n",
    "        vec = [0] * (len(categories) - 1)\n",
    "        category_num = category_to_int_dict[category]\n",
    "        if category_num > 0:\n",
    "            vec[category_num-1] = 1\n",
    "        return np.array(vec)\n",
    "    \n",
    "    assert len(X[0]) == len(X[1]) and len(X)==2\n",
    "    categories = list(set(list(X[0]) + list(X[1])))\n",
    "    category_to_int_dict = {c:categories.index(c) for c in categories}\n",
    "    \n",
    "    dummies = []\n",
    "    for i in range(len(X[0])):\n",
    "        dummies.append(_cat_to_vec(X[0,i], category_to_int_dict) + _cat_to_vec(X[1,i], category_to_int_dict))\n",
    "    return np.transpose(dummies)\n",
    "\n",
    "def predict_holdout(gene):\n",
    "    try:\n",
    "        exppheno = expression_phenotypes.loc[gene].values\n",
    "\n",
    "        # Select cis-snps and corresponding dosages\n",
    "        gene_annotation = gene_annotations.loc[gene]\n",
    "        gene_chr = gene_annotation[\"chr\"]\n",
    "        gene_start, gene_end = int(gene_annotation[\"start\"]), int(gene_annotation[\"end\"])\n",
    "        ciswindow_start, ciswindow_end = gene_start - ciswindow_size, gene_end + ciswindow_size\n",
    "        snp_annotations_chr = snp_annotations_by_chr[gene_chr]\n",
    "        cissnp_annotations = snp_annotations_chr[snp_annotations_chr[\"pos\"].between(ciswindow_start, ciswindow_end)]\n",
    "        cissnps = list(cissnp_annotations[\"varID\"])\n",
    "        # Extract cissnp dosages\n",
    "        cissnps_subset = list(set(cissnps).intersection(list(dosages.index)))\n",
    "        cisgenos = dosages.loc[cissnps_subset,:] # TODO: Ensure all cissnps are in cisgenos (necessary?)\n",
    "        # Extract cishaplo clusterings\n",
    "        cissnps_haplo = []\n",
    "        for cissnp in cissnps: cissnps_haplo += [cissnp+\"_h0\", cissnp+\"_h1\"]\n",
    "        cissnps_haplo = [c for c in cissnps_haplo if c in clusterings.index]\n",
    "        cisclusterings = clusterings.loc[cissnps_haplo,:]\n",
    "\n",
    "        # Skip this gene if there are fewer than two variants\n",
    "        if len(cisgenos) < 2:\n",
    "            return [gene] + [\"NA\"]*(dosages.shape[1]//10), (gene, \"NA\", \"NA\", 0, 0, \"CS<2\")\n",
    "\n",
    "        # Encode categorical clusterings with dummy variables\n",
    "        for _ in range(0, len(cissnps_haplo), 2):    \n",
    "            dummies = get_haplotype_cluster_dummies(cisclusterings.iloc[0:2,:].values)\n",
    "            dummies_index = [cisclusterings.iloc[0:2,:].index[0].split(\"h0\")[0]+\"d\"+str(i) for i in range(dummies.shape[0])]\n",
    "            dummies_df = pd.DataFrame(dummies, index=dummies_index, columns=cisclusterings.columns)\n",
    "            cisclusterings = cisclusterings.append(dummies_df)\n",
    "            cisclusterings.drop(cisclusterings.index[0:2], axis=0, inplace=True)\n",
    "\n",
    "        cisgenos = cisgenos.append(cisclusterings).T\n",
    "\n",
    "        # Separate cisgenos and expphenos into train and test sets\n",
    "        cisgenos_train = cisgenos.drop(cisgenos.index[::10], axis=0)\n",
    "        cisgenos_test = cisgenos.loc[cisgenos.index[::10]]\n",
    "        exppheno_train = [Y for i,Y in enumerate(exppheno) if i%10 != 0]\n",
    "        exppheno_test = [Y for i,Y in enumerate(exppheno) if i%10 == 0]\n",
    "\n",
    "        # Fit model for this gene\n",
    "        model = ElasticNetCV(l1_ratio=0.5, n_alphas=n_alphas, cv=n_folds, tol=tol, random_state=SEED, n_jobs=1)\n",
    "        model.fit(cisgenos_train.values, exppheno_train)\n",
    "\n",
    "        if all(coef==0 for coef in model.coef_): # NOTE: Near-zero betas are not captured here\n",
    "            r_train = \"NA\"\n",
    "            return [gene] + [\"NA\"]*(dosages.shape[1]//10), (gene, \"NA\", \"NA\", 0, 0, \"NB\")\n",
    "        \n",
    "        # Save regression coefficients\n",
    "        pd.DataFrame(model.coef_, index=cisgenos.columns).T.to_csv(\"output/model{}/betas/{}/{}.txt\".format(MODEL_NUM, tissue, gene),\n",
    "                                                                   sep=\" \", index=False)\n",
    "        n_dosage_nonzero_betas = np.count_nonzero(model.coef_[:-len(cisclusterings)])\n",
    "        n_clustering_nonzero_betas = np.count_nonzero(model.coef_[-len(cisclusterings):])\n",
    "        \n",
    "        r_train = pearsonr(exppheno_train, model.predict(cisgenos_train.values))[0]\n",
    "\n",
    "        exppheno_pred = model.predict(cisgenos_test.values)\n",
    "        r_test = pearsonr(exppheno_test, exppheno_pred)[0]\n",
    "\n",
    "        return [gene] + list(exppheno_pred), (gene, r_train, r_test, n_dosage_nonzero_betas, n_clustering_nonzero_betas, \"OK\")\n",
    "    except:\n",
    "        print(\"{}: unknown error\".format(gene))\n",
    "        return [gene] + [\"NA\"]*(dosages.shape[1]//10), (gene, \"NA\", \"NA\", \"NA\", \"NA\", \"ERR\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run model in parallel across genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(\"output/model{}/betas/{}/\".format(MODEL_NUM, tissue), exist_ok=True)\n",
    "\n",
    "genes = list(expression_phenotypes.index)\n",
    "\n",
    "model_outputs = []\n",
    "for gene in genes:\n",
    "    model_outputs.append(predict_holdout(gene))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save model predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ind1</th>\n",
       "      <th>Ind11</th>\n",
       "      <th>Ind21</th>\n",
       "      <th>Ind31</th>\n",
       "      <th>Ind41</th>\n",
       "      <th>Ind51</th>\n",
       "      <th>Ind61</th>\n",
       "      <th>Ind71</th>\n",
       "      <th>Ind81</th>\n",
       "      <th>Ind91</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gene000</th>\n",
       "      <td>0.0157724</td>\n",
       "      <td>0.0157724</td>\n",
       "      <td>0.0157724</td>\n",
       "      <td>0.0157724</td>\n",
       "      <td>0.0157724</td>\n",
       "      <td>0.0157724</td>\n",
       "      <td>0.0157724</td>\n",
       "      <td>0.0157724</td>\n",
       "      <td>0.0157724</td>\n",
       "      <td>0.0157724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene001</th>\n",
       "      <td>-0.0757439</td>\n",
       "      <td>-0.0757439</td>\n",
       "      <td>-0.0757439</td>\n",
       "      <td>-0.0757439</td>\n",
       "      <td>-0.0757439</td>\n",
       "      <td>-0.0757439</td>\n",
       "      <td>-0.0757439</td>\n",
       "      <td>-0.0757439</td>\n",
       "      <td>-0.0757439</td>\n",
       "      <td>-0.0757439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene002</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene003</th>\n",
       "      <td>-0.315475</td>\n",
       "      <td>-0.297647</td>\n",
       "      <td>-0.252577</td>\n",
       "      <td>-0.307353</td>\n",
       "      <td>0.258391</td>\n",
       "      <td>-0.466567</td>\n",
       "      <td>-0.463559</td>\n",
       "      <td>-0.327009</td>\n",
       "      <td>-0.0571043</td>\n",
       "      <td>-0.571866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene004</th>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Ind1      Ind11      Ind21      Ind31      Ind41      Ind51  \\\n",
       "gene_id                                                                     \n",
       "Gene000  0.0157724  0.0157724  0.0157724  0.0157724  0.0157724  0.0157724   \n",
       "Gene001 -0.0757439 -0.0757439 -0.0757439 -0.0757439 -0.0757439 -0.0757439   \n",
       "Gene002         NA         NA         NA         NA         NA         NA   \n",
       "Gene003  -0.315475  -0.297647  -0.252577  -0.307353   0.258391  -0.466567   \n",
       "Gene004         NA         NA         NA         NA         NA         NA   \n",
       "\n",
       "             Ind61      Ind71      Ind81      Ind91  \n",
       "gene_id                                              \n",
       "Gene000  0.0157724  0.0157724  0.0157724  0.0157724  \n",
       "Gene001 -0.0757439 -0.0757439 -0.0757439 -0.0757439  \n",
       "Gene002         NA         NA         NA         NA  \n",
       "Gene003  -0.463559  -0.327009 -0.0571043  -0.571866  \n",
       "Gene004         NA         NA         NA         NA  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "expression_phenotypes_pred_list = [o[0] for o in model_outputs]\n",
    "expression_phenotypes_pred = pd.DataFrame(expression_phenotypes_pred_list, columns=[\"gene_id\"] + list(dosages.columns[::10])).set_index(\"gene_id\")\n",
    "display(expression_phenotypes_pred.head())\n",
    "fname = \"output/model{}/{}_pred.txt\".format(MODEL_NUM, tissue)\n",
    "os.makedirs(os.path.dirname(fname), exist_ok=True)\n",
    "expression_phenotypes_pred.to_csv(fname, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
